### DISCUSSION TITLE
Difference between kvrocks and pika

### DISCUSSION THREAD

**Discussion by dongdongwcpp** [apache/kvrocks#319] [Q&A] [answered]
What is the difference between pika and kvrocks? Does kvrocks provide better performance than pika?
Any improvements on the blackwidow storage engine design？

**Comment by git-hulk** [comment, marked as answer]
It's a good question, many users should have the same confusion, I would try to explain the differences between them. To be noticed, it only represents my perspective and can't promise fairness.

**From the key design and features:**

* We prefer to keep consistent with Redis semantics which the Pika didn't. e.g.  Pika allows different types to have the same key which wasn't expected for most users.
* We do much work on improving the performance and decreasing the latency. Like auto-adjusting the block and SST with the key-value size, as well as using the compaction checker to reduce the full compaction on the RocksDB.
* More features were supported by Kvrocks, like Lua and Transaction
* We also care about profiling which would help users to find the performance issue. Users can use the `perf` command to investigate the issue when suffering the latency spike.
* Finally, the core competitiveness of the kvrocks is simple and maintainable enough for most users. The tendis should be too complex for most users/companies, and also hard to deploy.

**From the community:**

Kvrocks has been denoted to the Apache, therefore it didn't belong to any companies or organizations. Also, core team members were from different companies with many years of experience in the distributed key-value field.

**Reply by dongdongwcpp** [reply]
thanks to reply, great job.
instead of seeking by O(N) to implement `ZSet::Range`，any plan to do some optimizations? I noticed Tendis use in memory skiplist to avoid seeking/seekprev,maybe that's a  good idea?

**Reply by git-hulk** [reply]
Thanks for your tips, I'm not sure how much the performance would be improved if we cache the key ourselves instead of using the rocksdb block cache. We can measure the improvement and determine whether to do that since it would bring much extra complexity.
